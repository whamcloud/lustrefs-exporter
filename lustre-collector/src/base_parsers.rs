// Copyright (c) 2021 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use combine::{
    attempt,
    error::{Format, ParseError},
    many1, one_of,
    parser::{
        char::{alpha_num, digit, newline, string},
        repeat::take_until,
    },
    stream::Stream,
    token, unexpected, unexpected_any, value, Parser,
};

use crate::types::{Param, Target};

pub(crate) fn period<I>() -> impl Parser<I, Output = char>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    token('.')
}

pub(crate) fn equals<I>() -> impl Parser<I, Output = char>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    token('=')
}

pub(crate) fn word<I>() -> impl Parser<I, Output = String>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    many1(alpha_num().or(token('_')))
}

/// Parses a target name
pub(crate) fn target<I>() -> impl Parser<I, Output = Target>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    many1(alpha_num().or(one_of("_-".chars()))).map(Target)
}

/// Takes many consecutive digits and
/// returns them as u64
pub(crate) fn digits<I>() -> impl Parser<I, Output = u64>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    many1(digit()).then(|x: String| match x.parse::<u64>() {
        Ok(n) => value(n).left(),
        Err(e) => unexpected_any(Format(e)).right(),
    })
}

pub(crate) fn till_newline<I>() -> impl Parser<I, Output = String>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    take_until(newline())
}

pub(crate) fn till_period<I>() -> impl Parser<I, Output = String>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    take_until(period())
}

pub(crate) fn string_to<I>(x: &'static str, y: &'static str) -> impl Parser<I, Output = String>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    string(x).map(move |_| String::from(y))
}

pub(crate) fn not_words<I>(xs: &'static [&'static str]) -> impl Parser<I, Output = String>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    attempt(word().then(move |y| {
        for &x in xs {
            if x == y {
                return unexpected(x).map(|_| "".to_string()).right();
            }
        }

        value(y).left()
    }))
}

pub(crate) fn param<I>(x: &'static str) -> impl Parser<I, Output = Param>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    attempt(string(x).skip(equals()))
        .map(|x| Param(x.to_string()))
        .message("while getting param")
}

pub(crate) fn param_period<I>(x: &'static str) -> impl Parser<I, Output = Param>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    attempt(string(x).skip(token('.')))
        .map(|x| Param(x.to_string()))
        .message("while getting param")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::Param;

    #[test]
    fn test_param() {
        let result = param("memused").parse("memused=77991501\n");

        assert_eq!(result, Ok((Param("memused".to_string()), "77991501\n")))
    }

    #[test]
    fn test_param_period() {
        let result = param_period("exports").parse("exports.1.2.3.130@o2ib.stats=Y\n");

        assert_eq!(
            result,
            Ok((Param("exports".to_string()), "1.2.3.130@o2ib.stats=Y\n"))
        )
    }
}
