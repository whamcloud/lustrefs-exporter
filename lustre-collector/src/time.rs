// Copyright (c) 2021 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::base_parsers::{digits, till_newline};
use combine::stream::Stream;
use combine::{
    attempt,
    parser::char::{spaces, string},
};
use combine::{error::ParseError, parser::char::newline};
use combine::{optional, token, Parser};

fn time<I>(name: &'static str) -> impl Parser<I, Output = String>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    (
        string(name).skip(optional(token(':'))),
        spaces(),
        digits().skip(token('.')),
        digits().skip(till_newline()),
    )
        .map(|(_, _, secs, nsecs)| format!("{secs}.{nsecs}"))
}

pub(crate) fn time_triple<I>() -> impl Parser<I, Output = String>
where
    I: Stream<Token = char>,
    I::Error: ParseError<I::Token, I::Range, I::Position>,
{
    (
        time("snapshot_time")
            .message("While parsing snapshot_time")
            .skip(newline()),
        optional(
            attempt(
                time("start_time")
                    .skip(newline())
                    .message("While parsing start_time"),
            )
            .and(
                time("elapsed_time")
                    .skip(newline())
                    .message("While parsing elapsed_time"),
            ),
        ),
    )
        .map(|(time, _)| time)
}

#[cfg(test)]
mod tests {
    use combine::EasyParser;
    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn test_time() {
        let x = r#"snapshot_time:         1534158712.738772898 (secs.nsecs)
"#;

        let result = time("snapshot_time").parse(x);

        assert_eq!(result, Ok(("1534158712.738772898".to_string(), "\n",)));
    }
    #[test]
    fn test_time_no_colon() {
        let x = r#"snapshot_time             1534769431.137892896 secs.nsecs
"#;

        let result = time("snapshot_time").parse(x);

        assert_eq!(result, Ok(("1534769431.137892896".to_string(), "\n")));
    }

    #[test]
    fn test_time_triple() {
        let x = r#"snapshot_time             1684948453.142852820 secs.nsecs
start_time                1684946875.504329012 secs.nsecs
elapsed_time              1577.638523808 secs.nsecs
"#;

        let result = time_triple().easy_parse(x).unwrap();

        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_time_triple_back_compat() {
        let x = r#"snapshot_time             1596728874.484750908 secs.nsecs
req_waittime              31280 samples [usec] 11 2695 5020274 1032267156

"#;

        let result = time_triple().easy_parse(x).unwrap();

        assert_debug_snapshot!(result);
    }
}
