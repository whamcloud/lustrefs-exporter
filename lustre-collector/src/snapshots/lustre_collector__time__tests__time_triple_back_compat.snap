---
source: src/time.rs
expression: result
---
(
    "1596728874.484750908",
    "req_waittime              31280 samples [usec] 11 2695 5020274 1032267156\n\n",
)
